### PR TITLE
NVLS - fused reduce-scatter + residual + rms-norm + all-gather kernel

### PR DESCRIPTION
**Pull request by sidsingh-nvidia**
This PR build on top of #1997 and fuses reduce-scatter, residual, rms-norm, and all-gather ops - which are 4 memory bound ops typically found one after the other in LLMs. 

To activate these kernels, use these args - `--transformer-impl inference_optimized --inference-fuse-tp-communication`.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [x] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [x] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by sidsingh-nvidia** [comment]
@kvareddy please review

File: megatron/core/transformer/transformer_block.py [hunk 1]
```diff
@@ -383,6 +383,24 @@ def build_layer(layer_spec, layer_number):
         else:
             self.final_layernorm = None  # Either this or nn.Identity
 
+        if self.config.inference_fuse_tp_communication:
+            for i in range(len(self.layers)):
+                if i > 0:
+                    self.layers[i].self_attention.linear_qkv.skip_norm_and_all_gather = True
+                self.layers[i].self_attention.linear_proj.next_layer_norm_weights = self.layers[
```
Comment (sidsingh-nvidia) [inline megatron/core/transformer/transformer_block.py]: getter/setter functions for at self.layers...

Comment (sidsingh-nvidia) [reply megatron/core/transformer/transformer_block.py]: put this in a function

File: megatron/core/transformer/transformer_layer.py [hunk 2]
```diff
@@ -520,10 +521,19 @@ def _forward_attention(
...
         nvtx_range_push(suffix="self_attn_bda")
-        with self.bias_dropout_add_exec_handler():
-            hidden_states = self.self_attn_bda(self.training, self.config.bias_dropout_fusion)(
-                attention_output_with_bias, residual, self.hidden_dropout
-            )
+        using_fused_tp_communication_kernel = (not self.training) and (
+            self.config.inference_fuse_tp_communication
+        )
+        if using_fused_tp_communication_kernel:
+            # In inference optimized transformer layer, there is no bias and dropout
```
Comment (sidsingh-nvidia) [inline megatron/core/transformer/transformer_layer.py:534]: might want to fuse bias in the future - add a todo.

File: megatron/core/transformer/attention.py [hunk 3]
```diff
@@ -985,7 +986,13 @@ def forward(
         # =================
 
         nvtx_range_push(suffix="linear_proj")
-        output, bias = self.linear_proj(core_attn_out)
+        using_inference_optimized_layer = (not self.training) and (
+            self.config.transformer_impl == "inference_optimized"
+        )
+        if using_inference_optimized_layer:
```
Comment (sidsingh-nvidia) [inline megatron/core/transformer/attention.py]: at the transformer layer level, can we call self.blah_blah.set_residual_input()?

Comment (sidsingh-nvidia) [reply megatron/core/transformer/attention.py]: we are trying to avoid changing the forward pass API.

File: megatron/core/transformer/attention.py [hunk 4]
```diff
@@ -703,6 +703,7 @@ def forward(
         sequence_len_offset: Optional[int] = None,
         *,
         inference_params: Optional[BaseInferenceContext] = None,
+        residual: Optional[Tensor] = None,
```
Comment (sidsingh-nvidia) [inline megatron/core/transformer/attention.py]: do not want to change training APIs....

File: megatron/core/transformer/attention.py (hunk 3 above)
Comment (sidsingh-nvidia) [reply megatron/core/transformer/attention.py]: if inference_optimzied:
     set_residual() ...

File: megatron/core/tensor_parallel/inference_layers.py [hunk 5]
```diff
@@ -90,7 +91,24 @@ def __init__(
                 config.sequence_parallel
             ), "--transformer-impl=inference_optimized requires --sequence-parallel"
 
-    def _all_gather(self, x: torch.Tensor) -> None:
+        # Boolean to be toggled externally for skipping norm and all-gather.
+        # This is used when enabling fused reduce-scatter + add + rms-norm + all-gather
+        # in tensor parallelism. In this case, the preceeding RowParallelLinear layer
+        # has already applied the rms-norm and all-gather.
+        self.skip_norm_and_all_gather = False
```
Comment (timmoon10) [inline megatron/core/tensor_parallel/inference_layers.py:98]: This design for indirectly moving the norm out of `LayerNormLinear` and into the previous layer is similar to this Transformer Engine PR: https://github.com/NVIDIA/TransformerEngine/pull/2212. It's the most direct way to implement this fusion, but it will be tricky to maintain because it makes the compute non-obvious. Especially in training (so not relevant to this module), it makes the forward and backward passes not line up, so there's a risk of autograd bugs.

Comment (sidsingh-nvidia) [reply megatron/core/tensor_parallel/inference_layers.py:98]: Yeah, I agree that this is a bit ugly and that is why we have currently housed it in the inference-only layers. Even for inference, this feature is currently experimental and is disabled by default, since it relies on a number of assumptions about the model architecture.
Do you have any suggestions as to how we can improve code quality?

**Comment by sidsingh-nvidia** [comment]
@kvareddy - ready for re-review

File: megatron/core/transformer/transformer_layer.py [hunk 6]
```diff
@@ -494,6 +494,15 @@ def _forward_attention(
         else:
             input_layernorm_output = self.input_layernorm(hidden_states)
 
+        using_fused_tp_communication_kernel = (not self.training) and (
```
Comment (jaredcasper) [inline megatron/core/transformer/transformer_layer.py]: This could get confusing when people are reading this code for training purposes and confused with training tp comm overlap, etc. Maybe `using_fused_tp_inference_kernel` (communication can be implicit to keep name from being unreasonably long?)

File: megatron/core/transformer/transformer_layer.py [hunk 7]
```diff
@@ -669,6 +699,66 @@ def sharded_state_dict(
             apply_prefix_mapping(sharded_state_dict, prefixed_map)
         return sharded_state_dict
 
+    def configure_fused_tp_communication(
```
Comment (jaredcasper) [inline megatron/core/transformer/transformer_layer.py]: probably want inference in this name as well.

**Review by jaredcasper — APPROVED** [review]
